### DISCUSSION TITLE
Missing videodev.ko in /lib/modules

### DISCUSSION THREAD

**Discussion by pergolafabio** [home-assistant/operating-system#1938] [General]
### Describe the issue you are experiencing

Hu, i'm trying to load videodev support, according to this file : "device-support.config", it should be enabled for compiling

`CONFIG_VIDEO_DEV=m`

But if i access the host, i dont see the .ko file for it? how can it be missing? how can i modprobe/insmod it now?

thnx for feedback

![image](https://user-images.githubusercontent.com/43251136/169774343-1551379e-9246-4e4f-a857-62b00cb4f00e.png)

PS: i'm running on an esxi system

just for test i created another alpine system on my esxi, and then i see the videodev.ko file listed there:

![image](https://user-images.githubusercontent.com/43251136/169774553-7d5a6013-e1ce-431c-8a17-54ee0c20e915.png)

thnx for feedback!!

### What operating system image do you use?

generic-x86-64 (Generic UEFI capable x86-64 systems)

### What version of Home Assistant Operating System is installed?

8.1

### Did you upgrade the Operating System.

No

### Steps to reproduce the issue

none

### Anything in the Supervisor logs that might be useful for us?

```txt
No
```

### Anything in the Host logs that might be useful for us?

```txt
No
```

**Comment by agners** [comment]
According to the full `.config` (see [config.txt](https://github.com/home-assistant/operating-system/files/8752868/config.txt)) the driver is compiled in. Not sure why it is not a module but built-in, but it shouldn't matter as the driver is enabled.

**Reply by agners** [reply]
Converted to a discussion as this is not really an issue with HAOS.

> i was able to compile the .ko file, but in order to insmod/modprobe it, it neet dependencies like "videodev"

What kernel version/sources/configuration did you build against? You cannot just compile a kernel module against some kernel version/configuration, and expect it to work on a different build. To build out-of-tree modules for a kernel, you need to have the kernel headers and configuration for *exactly* the target kernel, however, HAOS is not designed to build kernel modules on the target system (there are no build tools or kernel headers etc. available).

If you want to add a kernel module to HAOS, you need to modify the buildroot build script sources. To read more about how HAOS is designed refer to [the developer website](https://developers.home-assistant.io/docs/operating-system/).

**Comment by pergolafabio** [comment]
Hi, thnx for feedback! But if I do a "modprobe video dev", I don't see it listed when I do an "lsmod" ?

**Reply by agners** [reply]
Because its not a kernel module. Only modules show up in `lsmod`.

**Comment by pergolafabio** [comment]
Ok, I need to dive in deeper :-) still learning...

But why do I see then an videodev.ko file on my alpine test system?

**Reply by agners** [reply]
Because Alpine decided to compile that driver as a kernel module.

**Comment by pergolafabio** [comment]
Also if I do a "modprobe videodev" on my alpine, I see it listed there with lsmod

**Reply by agners** [reply]
Same, Alpine decided to compile that driver as a kernel module, so it actually appears in `lsmod`.

**Comment by pergolafabio** [comment]
Ok, that makes sense :-)

So I don't need todo anything? I have another module that I need to load, and "videodev" is a dependency , but it fails loading because "videodev" is not loaded :-(

**Reply by agners** [reply]
> I have another module that I need to load, and "videodev" is a dependency

See, that is the problem: Your "other module" is built against the Alpine kernel. So the module assumes that videodev is a kernel module, (because it is built for the Alpine kernel). In Linux you cannot just build a module somewhere, and  expect it to load in another kernel.

You need to build your kernel module against *exaclty* the same sources, configuration and ideally toolchain, as the target kernel has been built with. And that is not really possible with HAOS. Hence you need to include the kernel module as part of HAOS itself.

**Comment by pergolafabio** [comment]
Ok, thnx for all feedback, appreciated!!

I'look for another way then :-)

**Comment by pergolafabio** [comment]
hey @agners 
ok ; i now compiled a buildroot, with exact same version of kernel 5.15.41 using qemu x64_x86 
the v4l2loopback page compiled fine, it generated me an .ko file ...

when i now do an "insmod v4lt2loopback.ko"  i get this error:
![image](https://user-images.githubusercontent.com/43251136/170051025-b25d7af8-bede-4bdb-8b02-6231251eb375.png)


indeed, a "modprobe videodev" was indeed not needed after all on buildroot :-)

any idea how i can troubleshoot that?

or is it possible to request an package in a next HassOS release with below?

```
BR2_PACKAGE_V4L2LOOPBACK=y
BR2_PACKAGE_V4L2LOOPBACK_UTILS=y
```

**Comment by pergolafabio** [comment]
i'm now testing with this tutorial : 

https://developers.home-assistant.io/docs/operating-system/getting-started/

and added in the OVA defconfig 

```
BR2_PACKAGE_V4L2LOOPBACK=y
BR2_PACKAGE_V4L2LOOPBACK_UTILS=y
```
lets see what happens if i take that generated  .ko file and copy it over

lets hope that works

if it should work, is it allowed to make a PR to include below in the def configs?

```
BR2_PACKAGE_V4L2LOOPBACK=y
BR2_PACKAGE_V4L2LOOPBACK_UTILS=y
```

**Reply by agners** [reply]
Hm, I see, `BR2_PACKAGE_V4L2LOOPBACK` builds an out-of-tree kernel module. If you enable that option, the generated image should include that kernel module. So you can just flash the new image of HAOS generated by your build.

> lets see what happens if i take that generated .ko file and copy it over

That probably will work, but you need to make sure to run a HAOS built from the same source tree (specifically, same version of kernel).

> if it should work, is it allowed to make a PR to include below in the def configs?

You probably only need `BR2_PACKAGE_V4L2LOOPBACK=y` in HAOS, as the utils will run in your add-on I assume? Using the utils from another source should work if the kernel <-> user space API is stable. For things in upstream Linux, the developer VERY much take care to not break the user space API. However, for out-of-tree kernel modules, its up to that particular project/developer, so YMMV...

That said, I am not a huge fan of integrating out-of-tree modules, as sometimes they fail to build and then I need to check whats wrong etc, so it adds maintance burden. But I am not strictly against it, especially if there is a good reason. What is your use case of this?

**Comment by pergolafabio** [comment]
Hi, yes, I succeeded indeed by copying over the .ko file... Utils indeed is not needed , also not for the addon we are creating... Thnx for help...
The loopback can create multiple /dev/video devices, that way we can inject for instance with ffmpeg an RTSP stream to it...

We are using the asterisk addon, from sip clients or Lovelace sip card, we can actually call now camera feeds...
But the actual usecase is for the people who are using asterisk, and a doorintercom (without camera) is when they receive a call from the door intercom, to inject the video to the call, so no videocam is needed anymore... Lots of users have intercoms without a camera... Now with the sofphone we can also see video from another source...

Is it possible to make a PR just for the loopback package only? Otherwise we need to recompile it again on every new haos release..

Thnx in advance!

**Reply by agners** [reply]
> Is it possible to make a PR just for the loopback package only? 

Sure, I'll add it for the next release.

**Comment by pergolafabio** [comment]
Perfect, and thnx for the all the help, much appreciated

**Comment by pergolafabio** [comment]
Can you link me the PR when you do the change? Then I can subscribe to it to know when it will be released :-)

**Comment by pergolafabio** [comment]
Hi @agners , was it added in new release 8.2?

**Comment by pergolafabio** [comment]
@agners , did you forget this to add? :-)
Can you make a PR for it?

Thnx much appreciated

**Reply by agners** [reply]
Sorry, thanks for the PR (https://github.com/home-assistant/operating-system/pull/2243).

**Comment by pergolafabio** [comment]
Np, I hope it's ok :-)
